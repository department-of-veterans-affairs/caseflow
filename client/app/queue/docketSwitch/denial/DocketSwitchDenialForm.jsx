import React from 'react';
import PropTypes from 'prop-types';

import { useForm } from 'react-hook-form';
import { yupResolver } from '@hookform/resolvers/yup';
import * as yup from 'yup';

import {
  DOCKET_SWITCH_DENIAL_TITLE,
  DOCKET_SWITCH_DENIAL_INSTRUCTIONS,
} from 'app/../COPY';
import AppSegment from '@department-of-veterans-affairs/caseflow-frontend-toolkit/components/AppSegment';
import Button from 'app/components/Button';
import { sprintf } from 'sprintf-js';
import DateSelector from 'app/components/DateSelector';
import TextareaField from 'app/components/TextareaField';
import { css } from 'glamor';

const schema = yup.object().shape({
  receiptDate: yup.date().required(),
  context: yup.string().required(),
});

export const DocketSwitchDenialForm = ({
  appellantName,
  onCancel,
  onSubmit,
}) => {
  const { register, handleSubmit, formState } = useForm({
    resolver: yupResolver(schema),
    mode: 'onChange',
  });

  const sectionStyle = css({ marginBottom: '24px' });

  return (
    <form
      className="docket-switch-denial"
      onSubmit={handleSubmit(onSubmit)}
      aria-label="Deny Docket Switch"
    >
      <AppSegment filledBackground>
        <h1>{sprintf(DOCKET_SWITCH_DENIAL_TITLE, appellantName)}</h1>
        <div {...sectionStyle}>{DOCKET_SWITCH_DENIAL_INSTRUCTIONS}</div>

        <DateSelector
          inputRef={register}
          type="date"
          name="receiptDate"
          label="What is the Receipt Date of the docket switch request?"
          strongLabel
        />

        <TextareaField
          inputRef={register}
          name="context"
          label="Provide context for this denial (this will be visible in the Case Timeline)"
          strongLabel
        />
      </AppSegment>
      <div className="controls cf-app-segment">
        <Button
          type="submit"
          name="submit"
          disabled={!formState.isValid}
          classNames={['cf-right-side']}
        >
          Confirm
        </Button>
        {onCancel && (
          <Button
            type="button"
            name="Cancel"
            classNames={['cf-right-side', 'usa-button-secondary']}
            onClick={onCancel}
            styling={{ style: { marginRight: '1rem' } }}
          >
            Cancel
          </Button>
        )}
      </div>
    </form>
  );
};

DocketSwitchDenialForm.propTypes = {
  appellantName: PropTypes.string.isRequired,
  onCancel: PropTypes.func,
  onSubmit: PropTypes.func.isRequired,
};
