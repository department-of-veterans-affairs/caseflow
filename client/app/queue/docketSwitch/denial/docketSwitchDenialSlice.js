import { createAsyncThunk } from '@reduxjs/toolkit';
import ApiUtil from '../../../util/ApiUtil';

export const completeDocketSwitchDenial = createAsyncThunk(
  'docketSwitch/deny',
  async (data) => {
    try {
      // Update this to conform to submission endpoint expectations
      const res = await ApiUtil.post('/docket_switches', { data });
      const result = res.body?.data;

      return result;
    } catch (error) {
      console.error('Error denying docket switch', error);
      throw error;
    }
  }
);
