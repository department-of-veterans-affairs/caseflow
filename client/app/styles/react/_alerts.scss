
.usa-alert {

  &.fixed {
    position: fixed;
    top: 0;
    margin-top: 0;
    z-index: 999;
    left: 50%;
    transform: translate(-50%, 0);
  }

  // Remove after USWDS V.2.0
  &.usa-alert-slim {
    background-size: 4rem;

    .usa-alert-body {
      padding-left: 3.5rem;
    }

    .usa-alert-text {
      font-size: 2rem;
    }
  }

  .usa-alert-text {
    line-height: 1.5;
  }
}
