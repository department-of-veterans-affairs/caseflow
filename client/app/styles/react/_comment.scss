.comment-control-button-container {
  display: flex;
  width: 100%;
  padding: 3px 0;
  justify-content: space-between;
}

.comment-save-button-container {
  display: inline-block;
  width: 100%;
  margin-top: 5px;
  margin-bottom: 17px;
}

.comment-control-button {
  padding: 0;

  &.horizontal {
    text-align: left;
  }
}

.comment-control-button-divider {
  padding: 0 10px;
}

.comment-container-selected {
  box-shadow: 3px 3px 5px $color-gray-light;
}

.comment-container {
  word-wrap: break-word;
  word-break: break-word;
  background-color: $color-white;
  padding: 15px;
  cursor: pointer;

  &:not(.comment-horizontal-container) {
    margin-bottom: 15px;
    border-radius: 3px;
    border: 1px solid $color-gray-lighter;
  }

  &.comment-horizontal-container {
    padding: 25px 25px 25px 40px;
  }
}

// Max width is enforced so the size of this comment box
// fits with existing textareas in style guide
.comment-size-container {
  max-width: 430px;

  .cf-form-textarea {
    margin-bottom: 20px;
  }
}

.comment-horizontal-container {
  display: inline-block;
  width: 100%;
}

.comment-horizontal-spacing {
  margin-right: 49px;
}

.commentIcon-container {
  position: absolute;
  z-index: 10;
  // These dimensions need to match the comment icon SVG dimensions.
  width: 40px;
  height: 40px;
}

