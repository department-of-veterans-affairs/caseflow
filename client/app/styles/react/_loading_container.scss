.loadingContainer-positioning {
  position: absolute;
  width: 100%;
  height: 100%;
}

.loadingContainer-table {
  width: 100%;
  height: 100%;
  display: table;
}

.loadingContainer-table-cell {
  text-align: center;
  display: table-cell;
  vertical-align: middle;
}

.loadingContainer-content {
  position: relative;
  z-index: 1;

  iframe {
    width: 100%;
    height: 80vh;

    &.form9Viewer {
      height: 500px;
    }
  }
}

.loadingContainer-container {
  position: relative;
  overflow: hidden;
}

.cf-loading-button-text {
  display: inline-block;
  margin-right: 10px;
  line-height: 0;
  position: relative;
}

.cf-loading-message {
  .loading-message-text {
    display: inline-block;
    font-style: italic;
    font-weight: bolder;
  }

}

.cf-loading-button-symbol {
  display: inline-block;
}
