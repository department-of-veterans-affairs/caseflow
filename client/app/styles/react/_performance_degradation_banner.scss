.usa-banner {
  background-color: $color-gold-lightest;
  text-align: center;
}

.usa-banner-inner {
  padding: .8em;
}

.banner-icon {
  display: inline-block;
  vertical-align: text-top;
  // height: 25px;
}

.banner-text {
  padding-left: 6px;
}
