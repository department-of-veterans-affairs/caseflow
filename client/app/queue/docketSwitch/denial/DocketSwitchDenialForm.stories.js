import React from 'react';

import { DocketSwitchDenialForm } from './DocketSwitchDenialForm';

const instructions = ['**Summary:** Summary\n\n**Is this a timely request:** Yes\n\n**Recommendation:** Grant all issues\n\n**Draft letter:** http://www.va.gov'];

export default {
  title: 'Queue/Docket Switch/DocketSwitchDenialForm',
  component: DocketSwitchDenialForm,
  decorators: [],
  parameters: {},
  args: {
    appellantName: 'Jane Doe',
    instructions,
  },
  argTypes: {
    onCancel: { action: 'cancel' },
    onSubmit: { action: 'submit' },
  },
};

const Template = (args) => <DocketSwitchDenialForm {...args} />;

export const Basic = Template.bind({});

Basic.parameters = {
  docs: {
    storyDescription:
      'Used by attorney in Clerk of the Board office to complete a denial of a docket switch request',
  },
};

