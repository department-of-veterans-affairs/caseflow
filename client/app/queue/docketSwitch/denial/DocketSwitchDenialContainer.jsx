import React from 'react';
import { useHistory, useParams } from 'react-router';
import { useDispatch, useSelector } from 'react-redux';
import { DocketSwitchDenialForm } from './DocketSwitchDenialForm';
import { completeDocketSwitchDenial } from './docketSwitchDenialSlice';
import { appealWithDetailSelector } from '../../selectors';
import {
  DOCKET_SWITCH_DENIAL_SUCCESS_TITLE,
  DOCKET_SWITCH_DENIAL_SUCCESS_MESSAGE,
} from 'app/../COPY';
import { sprintf } from 'sprintf-js';
import { showSuccessMessage } from '../../uiReducer/uiActions';

export const DocketSwitchDenialContainer = () => {
  const { appealId, taskId } = useParams();
  const { goBack, push } = useHistory();
  const dispatch = useDispatch();

  const appeal = useSelector((state) =>
    appealWithDetailSelector(state, { appealId })
  );

  const handleCancel = () => goBack();
  const handleSubmit = async (formData) => {

    const docketSwitch = {
      disposition: 'denied',
      receipt_date: formData.receiptDate,
      context: formData.context,
      task_id: taskId,
      old_docket_stream_id: appeal.id
    };

    const successMessage = {
      title: sprintf(DOCKET_SWITCH_DENIAL_SUCCESS_TITLE, appeal.appellantFullName),
      detail: DOCKET_SWITCH_DENIAL_SUCCESS_MESSAGE,
    };

    try {
      await dispatch(completeDocketSwitchDenial(docketSwitch));

      dispatch(showSuccessMessage(successMessage));
      push(`/queue/appeals/${appealId}`);
    } catch (error) {
      // Perhaps show an alert that indicates error, advise trying again...?
      console.error('Error Denying Docket Switch', error);
    }
  };

  return <DocketSwitchDenialForm
    appellantName={appeal.appellantFullName}
    onCancel={handleCancel}
    onSubmit={handleSubmit}
  />;
};
