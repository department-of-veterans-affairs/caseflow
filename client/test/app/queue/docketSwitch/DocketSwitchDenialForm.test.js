import React from 'react';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DocketSwitchDenialForm } from 'app/queue/docketSwitch/denial/DocketSwitchDenialForm';
import {
  DOCKET_SWITCH_DENIAL_TITLE,
  DOCKET_SWITCH_DENIAL_INSTRUCTIONS,
} from 'app/../COPY';
import { sprintf } from 'sprintf-js';

const instructions = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit';

describe('DocketSwitchDenialForm', () => {
  const onSubmit = jest.fn();
  const onCancel = jest.fn();
  const appellantName = 'Claimant 1';
  const defaults = { onSubmit, onCancel, appellantName };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    const { container } = render(<DocketSwitchDenialForm {...defaults} />);

    expect(container).toMatchSnapshot();

    expect(
      screen.getByText(sprintf(DOCKET_SWITCH_DENIAL_TITLE, appellantName))
    ).toBeInTheDocument();
    expect(
      screen.getByText(DOCKET_SWITCH_DENIAL_INSTRUCTIONS)
    ).toBeInTheDocument();
  });

  it('fires onCancel', async () => {
    render(<DocketSwitchDenialForm {...defaults} />);
    expect(onCancel).not.toHaveBeenCalled();

    await userEvent.click(screen.getByRole('button', { name: /cancel/i }));
    expect(onCancel).toHaveBeenCalled();
  });

  describe('form validation & submission', () => {
    const receiptDate = '2020-10-01';
    const fillForm = async () => {
      //   Set receipt date
      await fireEvent.change(screen.getByLabelText(/receipt date/i), {
        target: { value: receiptDate },
      });

      //   Enter context/instructions
      await userEvent.type(
        screen.getByRole('textbox', { name: /context/i }),
        instructions
      );
    };

    it('disables submit until all fields valid', async () => {
      render(<DocketSwitchDenialForm {...defaults} />);

      const submit = screen.getByRole('button', { name: /confirm/i });

      expect(onSubmit).not.toHaveBeenCalled();

      await waitFor(() => {
        expect(submit).toBeDisabled();
      });

      await userEvent.click(submit);
      expect(onSubmit).not.toHaveBeenCalled();

      await fillForm();

      await waitFor(() => {
        expect(submit).toBeEnabled();
      });

      await userEvent.click(submit);
      await waitFor(() => {
        expect(onSubmit).toHaveBeenCalled();
      });
    });

    it('fires onSubmit with correct values', async () => {
      render(<DocketSwitchDenialForm {...defaults} />);

      const submit = screen.getByRole('button', { name: /confirm/i });

      await fillForm();

      await userEvent.click(submit);

      waitFor(() => {
        expect(onSubmit).toHaveBeenCalledWith({
          receiptDate,
          context,
        });
      });
    });
  });
});
