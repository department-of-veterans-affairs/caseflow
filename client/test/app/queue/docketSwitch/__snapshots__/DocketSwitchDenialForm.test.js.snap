// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocketSwitchDenialForm renders correctly 1`] = `
<div>
  <form
    aria-label="Deny Docket Switch"
    class="docket-switch-denial"
  >
    <section
      class="cf-app-segment cf-app-segment--alt"
    >
      <h1>
        Switch Docket: Deny Claimant 1's Request
      </h1>
      <div
        data-css-1sbm5ab=""
      >
        Add the date and add any context for the denial of this switch.
      </div>
      <div
        class="cf-form-textinput   "
      >
        <label
          for="receiptDate"
        >
          <strong>
            <span>
              What is the Receipt Date of the docket switch request?
            </span>
          </strong>
        </label>
        <input
          class="cf-form-textinput"
          id="receiptDate"
          max="9999-12-31"
          name="receiptDate"
          placeholder="mm/dd/yyyy"
          type="date"
          value=""
        />
      </div>
      <div
        class="cf-form-textarea"
      >
        <label
          class="question-label"
          for="context"
        >
          <strong>
            <span>
              Provide context for this denial (this will be visible in the Case Timeline)
            </span>
          </strong>
        </label>
        <textarea
          id="context"
          name="context"
        />
      </div>
    </section>
    <div
      class="controls cf-app-segment"
    >
      <span>
        <button
          class="cf-right-side usa-button"
          id="submit-submit"
          type="submit"
        >
          Confirm
        </button>
      </span>
      <span>
        <button
          class="cf-right-side usa-button-secondary usa-button"
          id="button-Cancel"
          style="margin-right: 1rem;"
          type="button"
        >
          Cancel
        </button>
      </span>
    </div>
  </form>
</div>
`;
